### PR TITLE
Add RawArgsObject in myCallData.compile, with reorder of the properties

### PR DESCRIPTION
**Pull request by PhilippeR26**
## Motivation and Resolution

Improvement :
- `CallData.compile` accepts as input the `RawArgArray` and `RawArgsObject` types, but `myCallData.compile` accepts only `RawArgArray`.
- As this method has access to the contract ABI, it should be possible to send it objects whose properties are not ordered (as in any regular JS/TS objects).

utils/calldata is modified to solve these 2 subjects.

## Usage related changes

- Users will be able to prepare their constructor calls with `myCallData.compile`, using RawArgsObject.
- It works for both Cairo 0 and 1.
- Objects (including Uint256) are reordered in conformity with abi.
- If user added properties that are not in ABI, they are ignored.
- If user forgot one property, the call to provider/node will not be done ; a clear and explicit message will be sent as error (Starknet error messages are not really explicit).
- If user add a property array_len before  an array (seen several times in Discord), the propery is ignored, and the call succeeds anyway. 

## Development related changes

- Test for Cairo 0 are in contract.test.ts
- Test for Cairo 1 are in cairo1.test.ts

## Hypothesis:

For Cairo 1, hypothesis have been taken for complex tuple.
It's expected to have this format :
```typescript 
((core::integer::u8, core::integer::u16), core::integer::u256)
```
or 
```typescript 
(core::integer::u256, test_constructor4::test_constructor4::TestConstructor4::Order2, core::array::Array::<core::integer::u128>, core::integer::u64)
```
Code has been created in accordance. 

## Checklist:

- [x] Performed a self-review of the code
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes in code (API docs will be generated automatically)
- [x] Updated the tests
- [x] All tests are passing

### CODE REVIEW

**Review by tabaktoni — CHANGES_REQUESTED** [review]

File: src/utils/calldata/index.ts [hunk 1]
```diff
@@ -77,13 +90,206 @@ export class CallData {
...
+      const orderedObject = orderObject(argsCalldata, abiMethod.inputs);
+      args = Object.values(orderedObject);
+      //   // validate array elements to abi
+      validateFields(abiMethod, args, this.structs);
+    }
+
     const argsIterator = args[Symbol.iterator]();
-    const { inputs } = this.abi.find((abi) => abi.name === method) as FunctionAbi;
-    return inputs.reduce(
+    return abiMethod.inputs.reduce(
```
[the remark this answers is not in the document]
Comment (tabaktoni) [reply src/utils/calldata/index.ts:292]: I think with inputs is fine ;)

File: src/utils/calldata/propertyOrder.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        // eslint-disable-next-line @typescript-eslint/no-use-before-define
+        return myArray.map((myElem) => orderTuple(myElem, { name: '0', type: typeInArray }));
+      case isTypeArray(typeInArray):
+        return myArray.map((myElem) => orderArray(myElem, { name: '0', type: typeInArray }));
+      default: // is an array of litterals
+        return myArray;
+    }
+  };
+
+  const orderTuple = (myObj: RawArgsObject, abiParam: AbiEntry): object => {
```
Comment (penovicp) [inline src/utils/calldata/propertyOrder.ts:125]: Similar suggestions as `orderStruct`.

File: src/utils/calldata/propertyOrder.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  isCairo1Type,
+  isLen,
+  isTypeArray,
+  isTypeStruct,
+  isTypeTuple,
+  isTypeUint256,
+} from './cairo';
+import extractTupleMemberTypes from './tuple';
+
+function errorU256(key: string) {
```
[the remark this answers is not in the document]
Comment (PhilippeR26) [reply src/utils/calldata/propertyOrder.ts:13]: Sorry, I don't understand your proposal. I understood that the `throw` command needs as parameter an object, and here you proposes a function.
Could you develop?

[the remark this answers is not in the document]
Comment (PhilippeR26) [reply src/utils/calldata/propertyOrder.ts:13]: Sorry, I don't understand your proposal. I understood that the `throw` command needs as parameter an object, and here you proposes a function.
Could you develop?

File: src/utils/calldata/propertyOrder.ts [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+              } else {
+                // object without 'low' & 'high'
+                errorU256(abiParam.name);
+              }
+            } else {
+              // BigNumberish --> just copy
+              copyProperty2struct();
+            }
+          }
+          break;
```
[the remark this answers is not in the document]
Comment (PhilippeR26) [reply src/utils/calldata/propertyOrder.ts:68]: I have modified with this spirit, but as `if` is a way to filter types, I can't have something so flatten.

File: src/utils/calldata/propertyOrder.ts [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+  throw Error(
+    `Your object includes the property : ${key}, containing an Uint256 object without the 'low' and 'high' keys.`
+  );
+}
+export default function orderPropsByAbi(
+  myInputObj: RawArgsObject,
+  abiOfObject: AbiEntry[],
+  structs: AbiStructs
+): object {
+  const orderStruct = (myObj: RawArgsObject, abiObject: AbiEntry[]): object => {
```
[the remark this answers is not in the document]
Comment (PhilippeR26) [reply src/utils/calldata/propertyOrder.ts:23]: Done.

**Comment by PhilippeR26** [comment]
Except `Error`  request of Petar, all is implemented.

File: src/utils/calldata/propertyOrder.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (penovicp) [reply src/utils/calldata/propertyOrder.ts:13]: The suggestion is to change the function from returning `void` and throwing internally to just returning an `Error`.

Function change:
```
function errorU256(key: string) {
  return Error(...);
}
```
Usage change:
```
// from
errorU256(abiParam.name);

// to
throw errorU256(abiParam.name);
```

The main benefit of this approach is that most code analysis tools disregard the internal throw even if it is guaranteed to always happen, for example placing `errorU256(...)` at the start of a method wouldn't generate an `unreachable code` warning for the rest of the method code while placing `throw errorU256(...)` would.

**Comment by PhilippeR26** [comment]
All is now implemented.

**Review by penovicp — APPROVED** [review]

**Pull request merged**